text/0000-tracked-properties.md

```diff
@@ -227,34 +248,350 @@ memoization) if a property shows up as a bottleneck during profiling.
+has not been updated to tracked properties. As such, to make the transition as
+easy as possible, `tracked` will also be made to work with the classic
+object-model:
+
+```js
+const Config = Service.extend({
+  polling: tracked({
+    value: {
+      shouldPoll: false,
+      pollInterval: -1,
+    }
+  }),
+
+  init() {
+    this._super(...arguments);
+
+    fetch('config/api/url')
+      .then(r => r.json())
+      .then(polling => set(this, 'polling', polling));
+  }
+})
+```
+
+Tracked getters and setters will also be allowed, with the `get` and `set` keys
+on the passed in configuration. This will allow existing libraries to transition
+incrementally, and add tracked support minimally where necessary.
+
+This form will _not_ be allowed on native classes, and will hard error if it is
+attempted. Additionally, default values will be defined on the _prototype_ to
+maintain consistency with the classic object-model.
 
 ## How we teach this
 
-> What names and terminology work best for these concepts and why? How is this
-idea best presented? As a continuation of existing Ember patterns, or as a
-wholly new one?
+There are two different aspects of tracked properties which need to be
+considered for the learning story:
+
+1. **General usage.** Which properties should I mark as tracked? How do I
+consume them? How do I trigger changes?
+2. **Interop with legacy systems.** How do I safely consume tracked properties
+from legacy classes and computeds? How do I safely consume legacy APIs from
+tracked properties?
+
+### General Usage
+
+The mental model with tracked properties is that anything _mutable_ should be
+tracked. If a value will ever change, it should have the `@tracked` decorator
+attached to it.
+
+After that, usage should be "Just Javascript". You can safely access values
+using any syntax you like, including desctructuring, and you can update values
+using standard assignments.
+
+```js
+// Dot notation
+const fullName = person.fullName;
+// Destructuring
+const { fullName } = person;
+// Bracket notation for computed property names
+const fullName = person['fullName'];
+
+// Simple assignment
+this.firstName = "Yehuda";
+// Addition assignment (+=)
+this.lastName += "Katz";
+// Increment operator
+this.age++;
+```
+
+#### Triggering Updates on Complex Objects
 
-> Would the acceptance of this proposal mean the Ember guides must be
-re-organized or altered? Does it change how Ember is taught to new users
-at any level?
+There may be cases where users want to update values in complex, untracked
+objects such as arrays or POJOs. `@tracked` will only be usable with class
+syntax at first, and while it may make sense to formalize these objects into
+tracked classes in some cases, this will not always be the case.
 
-> How should this feature be introduced and taught to existing Ember
-users?
+To do this, users can re-set a tracked value directly after its inner values
+have been updated.
```

Comment:
This is at odds with the development-mode error that you get when mutating an untracked property, as described above. Given that error, I think this only applies to array mutations or wholesale object replacement; you cannot mutate an object's property and then re-set its tracked value.

Reply:
Ah, that's a good point. Yeah I'll update this to note that it is just for arrays